share/man/man4/it.4
.\"	$NetBSD: it.4,v 1.5 2007/07/05 16:00:39 xtraeme Exp $
.\"     $OpenBSD: it.4,v 1.7 2006/03/29 14:10:51 jsg Exp $
.\"
.\" Copyright (c) 2006-2007 Juan Romero Pardines <juan@xtrarom.org>
.\" Copyright (c) 2003 Julien Bordet <zejames@greygats.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd July 4, 2007
.Dt IT 4
.Os
.Sh NAME
.Nm it
.Nd ITE IT8705F, IT8712F, IT8716F and SiS950 hardware monitors
.Sh SYNOPSIS
.Cd "it0 at isa? port 0x290"
.Sh DESCRIPTION
The
.Nm
driver provides support for the
.Tn IT8705F , IT8712F , IT8716F
and
.Tn SiS950
hardware monitors to be used with the
.Xr envsys 4
interface.
.Pp
The
.Nm
driver has 15 sensors:
.Bl -column "Sensor" "Units" "Typical" -offset indent
.It Sy "Sensor" Ta Sy "Units" Ta Sy "Typical Use"
.It Li "Temp0" Ta "uK" Ta "CPU Temp"
.It Li "Temp1" Ta "uK" Ta "System Temp"
.It Li "Temp2" Ta "uK" Ta "Aux Temp"
.It Li "VIN0" Ta "uV DC" Ta "Vcore A"
.It Li "VIN1" Ta "uV DC" Ta "Vcore B"
.It Li "VIN2" Ta "uV DC" Ta "+3.3V"
.It Li "VIN3" Ta "uV DC" Ta "+5V"
.It Li "VIN4" Ta "uV DC" Ta "+12V"
.It Li "VIN5" Ta "uV DC" Ta "-12V"
.It Li "VIN6" Ta "uV DC" Ta "-5V"
.It Li "VIN7" Ta "uV DC" Ta "STANDBY"
.It Li "VBAT" Ta "uV DC" Ta "VBAT"
.It Li "Fan0" Ta "RPM" Ta "CPU Fan"
.It Li "Fan1" Ta "RPM" Ta "System Fan"
.It Li "Fan2" Ta "RPM" Ta "Aux Fan"
.El
.Sh SEE ALSO
.Xr envsys 4 ,
.Xr envstat 8
.Sh NOTES
By default the
.Nm
driver uses the port address
.Tn 0x290
but another manufacturers are known to use
.Tn 0xc00 or 0xd00
port addresses.
.Sh HISTORY
The
.Nm
driver first appeared in
.Ox 3.4
and then it was ported to
.Nx 4.0 .
.Sh AUTHORS
.An -nosplit
The
.Nm
driver was written by
.An Julien Bordet Aq zejames@greyhats.org
and was ported to
.Nx
by
.An Juan Romero Pardines Aq juan@xtrarom.org .
.Sh BUGS
Interrupt support is unimplemented.


doc/CHANGES
LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.889 $>


[Note: This file does not mention every change made to the NetBSD source tree.
For an exhaustive list, please subscribe to the `source-changes' mailing list,
or see the mailing list archives.  For more information, send a message
containing just the word `help' to `majordomo@NetBSD.org'.]


Changes from NetBSD 4.0 to NetBSD 5.0:
	evbppc: add support for Xilinx Virtex II-Pro/4-FX. [freza 20061202]
	mac68k: add support for four more Nubus video cards, including the 
		VillageTronic MacPicasso 340. [hauke 20061206]
	ixp425: add support for NPE Ethernet. [scw 20061210]
	postfix(1): Updated to 2.3.5. [rpaulo 20061220]
	nfs: Remove NQNFS support from kernel. [yamt 20061227]
	mount_psshfs(8): add puffs sshfs [pooka 20061229]
  	zaurus: add support for Sharp C3x00 PDA. [ober 20070105]
	i386: add support for Microsoft Xbox. [jmcneill 20070105]
	ntp: Import ntp 4.2.4 (20061228). Tracking of network interface address
		changes is now supported - no more ntpd restarts when interface
		addresses are re-configured (WLAN/DSL/ppp/...). [kardel 20070106]
	kernel: global 'securelevel' variable removed; it is now internal to
		the secmodel. [elad 20070110]
	GCC: Updated to version 4.1.2-20070110.  [mrg 20070111]
	acpitools: added from FreeBSD via fvdl [christos 20070114]
	scsictl(8): added setspeed command, to set the highest speed that the
		optical drive should use for reading data,
		contributed by Kimura Fuyuki. [bouyer 20070122]
	i386, amd64: add cmos(4), driver for x86 CMOS RAM. [dyoung 20070206]
	newlock2: Merge the newlock2 branch. Introduces a number of new
		kernel synchronization primitives, improves scalability
		on MP systems, and replaces the existing SA threads model
		with a 1:1 threading model. [ad 20070209]
	news68k: Implement generic softintr(9) support. [tsutsui 20070210]
	refuse: added file system in userspace compatibility functionality,
		layered on top of libpuffs [agc 20070211]
	i386: added an opencrypto provider for VIA ACE (AES encryption
		instructions), ported from OpenBSD. [daniel 20070217]
	sdiff(1): Replaced by OpenBSD sdiff(1) (public domain). [rmind 20070219]
	btuart(4): Add support for Bluetooth HCI UART (H4) driver.
		[kiyohara 20070221]
	btuartd(8): Add support for Bluetooth HCI UART (H4) daemon.
		[kiyohara 20070221]
        mutex(9), rwlock(9): Support priority inheritance. [yamt 20070226]
	eeprom(8): Add the missing kernel bits on arm and powerpc that kept
		eeprom from actually changing firmware settings
		[macallan 20070228]
	MAKEDEV(8), MAKEDEV.local(8): Let MAKEDEV.local use functions
		defined in MAKEDEV. [apb 20070303]
	x68k: Add generic softintr(9) support. [tsutsui 20070304]
	luna68k: Add generic softintr(9) support. [tsutsui 20070304]
	file(1): update to 4.20.  [pooka 20070304]
	hfs: Add read-only Apple HFS+ file system support. [dillo 20070305]
	i386, amd64: switch bootable iso images to native boot format, and add
		different bootable kernels on the iso images. Enable ACPI
		in default INSTALL and GENERIC kernels. [bouyer 20070306]
	openssl: Import OpenSSL 0.9.8e. [mjf 20070307]
	ipv6: add ipv6 fast forward. [liamjfoy 20070307]
	openssh: import 4.6 [christos 20070310]
	aiboost(4): ASUS AI Booster ACPI Hardware monitor driver,
		adapted from FreeBSD for envsys(4). [xtraeme 20070313]
	evbmips: add Infineon ADM5120 port. [dyoung 20070320]
	i386, amd64: Intel(R) On Demand Clock Modulation driver, to save
		power and overheating in CPUs supporting the Thermal
		Monitor feature (TM). Adapted from OpenBSD/FreeBSD.
		See options(4) for more information. [xtraeme 20070320]
	libc: add support for LC_TIME [manu 20070328]
	sh5, evbsh5: Port deleted [scw 20070408]
	genfb: generic framebuffer console driver with PCI and SBus
		frontends [macallan 20070411]
	sgimips: add support for SGI O2 PS/2 controller macekbc(4) and
		onboard display adapter crmfb(4) [jmcneill 20070412]
	sgimips: import mavb(4) audio device driver for SGI O2 from
		OpenBSD [jmcneill 20070412]
	sparc64: add support for hardware monitoring on
		Ultra Enterprise 450. See envctrl(4) [tnn 20070414]
	coda: overhaul kernel module (vnode locking rules, uvm_fault
		handling, general cleanup) [gdt 20070415]
	puffs(4): Support NFS exporting puffs file servers [pooka 20070416]
	xargs(1): Bring in xargs from FreeBSD to gain -I, -J etc. but keep our
		GNU compatible exit values. [christos 20070418]
	bluetooth(4): add 'service level' security [plunky 20070421]
	mount_9p(8): Add support to mount a 9P file service with puffs.
		[pooka 20070421]
	kernel: Remove the CHECK_ALT_xxx() defines and emul_find() used
		to locate files within an emulation's root directory.
		Instead get namei() to check the emulation root first.
		[dsl 20070422]
	prep: add support for the MTX604 RAVEN based machine.  Other RAVEN
		based machines are also likely to work now. [garbled 20070426]
	kernel: Added support for POSIX Asynchronous I/O. [rmind 20070430]
	puffs(3): Add puffs_framebuf(3), which abstracts the buffer
		management and event loop for networked file systems
		using puffs_cc(3).  Convert mount_psshfs(8) and 
		mount_9p(8) to use it.  [pooka 20070505]
	kernel: Add the TFTPROOT kernel option for TFTP'ing root RAMdisk 
		at root mount time. [manu 20070508]
	pf(4): Make 'nat' and 'rdr' translation rules obey state
		policy flags ('if-bound', 'group-bound', 'floating').
		Extend pf.conf(5) syntax to express the translation-state
		policies. [dyoung 20070510]
	kernel: add jmide(4), a driver for the JMicron Technology JMB36x
		PCIe to SATA II/PATA controllers. [bouyer 20070515]
	idlelwp: yamt-idlelwp branch was merged.
		1. separate context switching and thread scheduling.
		2. implement idle lwp.
		3. clean up related MD/MI interfaces.
		4. make scheduler(s) modular.
		[yamt 20070518]
	ld.elf_so: Support Solaris $ORIGIN etc. expansions in paths.
		[christos 20070518]
	postfix(1): Updated to 2.4.1 [heas 20070519]
	uslsa(4): added CP210x USB-RS232 ucom driver, as submitted by

sys/dev/isa/it.c
/*	$NetBSD: it.c,v 1.7.8.1 2007/07/11 20:06:29 mjf Exp $	*/
/*	$OpenBSD: it.c,v 1.19 2006/04/10 00:57:54 deraadt Exp $	*/

/*
 * Copyright (c) 2006-2007 Juan Romero Pardines <juan@xtrarom.org>
 * Copyright (c) 2003 Julien Bordet <zejames@greyhats.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITD TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITD TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Driver for the iTE IT8705/IT871[26]F Super I/O hardware monitor.
 */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD: it.c,v 1.7.8.1 2007/07/11 20:06:29 mjf Exp $");

#include <sys/param.h>
#include <sys/systm.h>
#include <sys/kernel.h>
#include <sys/proc.h>
#include <sys/device.h>
#include <sys/malloc.h>
#include <sys/errno.h>
#include <sys/conf.h>
#include <sys/envsys.h>
#include <sys/time.h>

#include <machine/bus.h>
#include <machine/intr.h>

#include <dev/isa/isareg.h>
#include <dev/isa/isavar.h>

#include <dev/sysmon/sysmonvar.h>

#include <dev/isa/itvar.h>

#define IT_VOLTSTART_IDX 	3 	/* voltage start index */
#define IT_FANSTART_IDX 	12 	/* fan start index */

#if defined(ITDEBUG)
#define DPRINTF(x)		do { printf x; } while (0)
#else
#define DPRINTF(x)
#endif

/*
 * IT87-compatible chips can typically measure voltages up to 4.096 V.
 * To measure higher voltages the input is attenuated with (external)
 * resistors.  Negative voltages are measured using a reference
 * voltage.  So we have to convert the sensor values back to real
 * voltages by applying the appropriate resistor factor.
 */
#define RFACT_NONE	10000
#define RFACT(x, y)	(RFACT_NONE * ((x) + (y)) / (y))

/* autoconf(9) functions */
static int  it_isa_match(struct device *, struct cfdata *, void *);
static void it_isa_attach(struct device *, struct device *, void *);

CFATTACH_DECL(it_isa, sizeof(struct it_softc),
    it_isa_match, it_isa_attach, NULL, NULL);

/* driver functions */
static int it_check(bus_space_tag_t, int);
static uint8_t it_readreg(struct it_softc *, int);
static void it_writereg(struct it_softc *, int, int);

/* envsys(9) glue */
static void it_setup_sensors(struct it_softc *);
static void it_refresh_temp(struct it_softc *, envsys_data_t *);
static void it_refresh_volts(struct it_softc *, envsys_data_t *);
static void it_refresh_fans(struct it_softc *, envsys_data_t *);
static int it_gtredata(struct sysmon_envsys *, envsys_data_t *);

/* rfact values for voltage sensors */
static const int it_vrfact[] = {
	RFACT_NONE,	/* VCORE_A	*/
	RFACT_NONE,	/* VCORE_B	*/
	RFACT_NONE,	/* +3.3V	*/
	RFACT(68, 100),	/* +5V 		*/
	RFACT(30, 10),	/* +12V 	*/
	RFACT(21, 10),	/* -12V 	*/
	RFACT(83, 20),	/* -5V 		*/
	RFACT(68, 100),	/* STANDBY	*/
	RFACT_NONE	/* VBAT		*/
};

static int
it_isa_match(struct device *parent, struct cfdata *match, void *aux)
{
	struct isa_attach_args *ia = aux;
	int rv = 0;

	/* Must supply an address */
	if (ia->ia_nio < 1)
		return 0;

	if (ISA_DIRECT_CONFIG(ia))
		return 0;

	if (ia->ia_io[0].ir_addr == ISA_UNKNOWN_PORT)
		return 0;

	rv = it_check(ia->ia_iot, ia->ia_io[0].ir_addr);

	if (rv) {
		ia->ia_nio = 1;
		ia->ia_io[0].ir_size = 8;
		ia->ia_niomem = 0;
		ia->ia_nirq = 0;
		ia->ia_ndrq = 0;
	}

	return rv;
}

static void
it_isa_attach(struct device *parent, struct device *self, void *aux)
{
	struct it_softc *sc = (struct it_softc *)self;
	struct isa_attach_args *ia = aux;
	int i;
	uint8_t cr;

	ia->ia_iot = sc->sc_iot;

	if (bus_space_map(sc->sc_iot, ia->ia_io[0].ir_addr, 8, 0,
	    &sc->sc_ioh)) {
		aprint_error(": can't map i/o space\n");
		return;
	}

	sc->sc_idr = it_readreg(sc, IT_COREID);
	if (sc->sc_idr == IT_REV_8712)
		aprint_normal(": IT871[26]F Hardware monitor\n");
	else {
		sc->sc_idr = it_readreg(sc, IT_VENDORID);
		if (sc->sc_idr == IT_REV_8705)
			aprint_normal(": IT8705F Hardware monitor\n");
		else
			aprint_normal(": iTE unknown vendor id (0x%x)\n",
			    sc->sc_idr);
	}

	/* Activate monitoring */
	cr = it_readreg(sc, IT_CONFIG);
	SET(cr, 0x01);
	it_writereg(sc, IT_CONFIG, cr);

#ifdef notyet
	/* Enable beep alarms */
	br = it_readreg(sc, IT_BEEPEER);
	SET(br, 0x02);	/* Voltage exceeds limit */
	SET(br, 0x04);	/* Temperature exceeds limit */
	it_writereg(sc, IT_BEEPEER, br);
#endif

	/* Initialize sensors */
	for (i = 0; i < IT_NUM_SENSORS; ++i) {
		sc->sc_data[i].sensor = i;
		sc->sc_data[i].state = ENVSYS_SVALID;
	}

	it_setup_sensors(sc);

	/*
	 * Hook into the system monitor.
	 */
	sc->sc_sysmon.sme_name = sc->sc_dev.dv_xname;
	sc->sc_sysmon.sme_sensor_data = sc->sc_data;
	sc->sc_sysmon.sme_cookie = sc;
	sc->sc_sysmon.sme_gtredata = it_gtredata;
	sc->sc_sysmon.sme_nsensors = IT_NUM_SENSORS;
	
	if (sysmon_envsys_register(&sc->sc_sysmon))
		printf("%s: unable to register with sysmon\n",
		    sc->sc_dev.dv_xname);

}

static int
it_check(bus_space_tag_t iot, int base)
{
	bus_space_handle_t ioh;
	int rv = 0;
	uint8_t cr0, cr1;
 
	if (bus_space_map(iot, base, 8, 0, &ioh))
		return 0;

	bus_space_write_1(iot, ioh, IT_ADDR, IT_RES48);
	cr0 = bus_space_read_1(iot, ioh, IT_DATA);
	bus_space_write_1(iot, ioh, IT_ADDR, IT_RES52);
	cr1 = bus_space_read_1(iot, ioh, IT_DATA);

	if (cr0 == IT_RES48_DEFAULT && cr1 == IT_RES52_DEFAULT)
		rv = 1;

	bus_space_unmap(iot, ioh, 8);
	return rv;
}

static uint8_t
it_readreg(struct it_softc *sc, int reg)
{
	bus_space_write_1(sc->sc_iot, sc->sc_ioh, IT_ADDR, reg);
	return bus_space_read_1(sc->sc_iot, sc->sc_ioh, IT_DATA);
}

static void
it_writereg(struct it_softc *sc, int reg, int val)
{
	bus_space_write_1(sc->sc_iot, sc->sc_ioh, IT_ADDR, reg);
	bus_space_write_1(sc->sc_iot, sc->sc_ioh, IT_DATA, val);
}

#define COPYDESCR(x, y)				\
	do {					\
		strlcpy((x), (y), sizeof(x));	\
	} while (0)

static void
it_setup_sensors(struct it_softc *sc)
{
	int i;

	/* temperatures */
	for (i = 0; i < IT_VOLTSTART_IDX; i++)
		sc->sc_data[i].units = ENVSYS_STEMP;

	COPYDESCR(sc->sc_data[0].desc, "CPU Temp");
	COPYDESCR(sc->sc_data[1].desc, "System Temp");
	COPYDESCR(sc->sc_data[2].desc, "Aux Temp");

	/* voltages */
	for (i = IT_VOLTSTART_IDX; i < IT_FANSTART_IDX; i++) {
		sc->sc_data[i].units = ENVSYS_SVOLTS_DC;
		sc->sc_data[i].rfact = ENVSYS_FCHANGERFACT;
	}

	COPYDESCR(sc->sc_data[3].desc, "VCORE_A");
	COPYDESCR(sc->sc_data[4].desc, "VCORE_B");
	COPYDESCR(sc->sc_data[5].desc, "+3.3V");
	COPYDESCR(sc->sc_data[6].desc, "+5V");
	COPYDESCR(sc->sc_data[7].desc, "+12V");
	COPYDESCR(sc->sc_data[8].desc, "-12V");
	COPYDESCR(sc->sc_data[9].desc, "-5V");
	COPYDESCR(sc->sc_data[10].desc, "STANDBY");
	COPYDESCR(sc->sc_data[11].desc, "VBAT");

	/* fans */
	for (i = IT_FANSTART_IDX; i < IT_NUM_SENSORS; i++)
		sc->sc_data[i].units = ENVSYS_SFANRPM;

	COPYDESCR(sc->sc_data[12].desc, "CPU Fan");
	COPYDESCR(sc->sc_data[13].desc, "System Fan");
	COPYDESCR(sc->sc_data[14].desc, "Aux Fan");
}
#undef COPYDESCR

static void
it_refresh_temp(struct it_softc *sc, envsys_data_t *edata)
{
	int sdata;

	sdata = it_readreg(sc, IT_SENSORTEMPBASE + edata->sensor);
	/* sensor is not connected or reporting invalid data */
	if (sdata == 0 || sdata >= 0xfa) {
		edata->state = ENVSYS_SINVALID;
		return;
	}

	DPRINTF(("%s: sdata[temp%d] 0x%x\n", __func__, edata->sensor, sdata));
	/* Convert temperature to uK */
	edata->value_cur = sdata * 1000000 + 273150000;
	edata->state = ENVSYS_SVALID;
}

static void
it_refresh_volts(struct it_softc *sc, envsys_data_t *edata)
{
	uint8_t vbatcr = 0;
	int i, sdata;

	i = edata->sensor - IT_VOLTSTART_IDX;

	sdata = it_readreg(sc, IT_SENSORVOLTBASE + i);
	/* not connected */
	if (sdata == 0) {
		edata->state = ENVSYS_SINVALID;
		return;
	}

	/* 
	 * update VBAT voltage reading every time we read it, to get
	 * latest value.
	 */
	if (i == 8) {
		vbatcr = it_readreg(sc, IT_CONFIG);
		SET(vbatcr, IT_UPDATEVBAT);
		it_writereg(sc, IT_CONFIG, vbatcr);
	}

	DPRINTF(("%s: sdata[volt%d] 0x%x\n", __func__, i, sdata));

	/* voltage returned as (mV << 4) */
	edata->value_cur = (sdata << 4);
	/* rfact is (factor * 10^4) */
	edata->value_cur *= it_vrfact[i];

	if (edata->rfact)
		edata->value_cur += edata->rfact;
	else
		edata->rfact = it_vrfact[i];

	/* division by 10 gets us back to uVDC */
	edata->value_cur /= 10;
	edata->state = ENVSYS_SVALID;
}

static void
it_refresh_fans(struct it_softc *sc, envsys_data_t *edata)
{
	int i, mode, sdata, divisor, odivisor, ndivisor;

	i = edata->sensor - IT_FANSTART_IDX;
	sdata = divisor = odivisor = ndivisor = 0;

#define FANDATA()	it_readreg(sc, IT_SENSORFANEXTBASE + i)

	mode = it_readreg(sc, IT_FAN16);
	if (sc->sc_idr == IT_REV_8705) {
		odivisor = ndivisor = divisor = it_readreg(sc, IT_FAN);
		divisor >>= 3;
		sdata = it_readreg(sc, IT_SENSORFANBASE + i);
		if (mode & (1 << i))
			sdata += (FANDATA() << 8);
		if (!(mode & (1 << i)) && sdata == 0xff) {
			edata->state = ENVSYS_SINVALID;
			if (i == 2)
				ndivisor |= 0x40;
			else if ((divisor & 7) != 7) {
				ndivisor &= ~(7 << (i * 3));
				ndivisor |= ((divisor + 1) & 7) << (i * 3);
			}
		} else {
			if (i == 2)
				divisor = divisor & 1 ? 3 : 1;
			if ((sdata << (divisor & 7)) == 0)
				edata->state = ENVSYS_SINVALID;
			else {
				edata->value_cur =
				    1350000 / (sdata << (divisor & 7));
				edata->state = ENVSYS_SVALID;
			}
		}
		DPRINTF(("%s: sdata[fan%d] 0x%x div: 0x%x\n", __func__,
		    i, sdata, divisor));
		if (ndivisor != odivisor)
			it_writereg(sc, IT_FAN, ndivisor);
	} else {
		/* IT8712F, IT8716F */
		sdata = it_readreg(sc, IT_SENSORFANBASE + i);
		if (mode & (1 << i)) /* 16-bit mode enabled */
			sdata += (FANDATA() << 8);
		edata->state = ENVSYS_SVALID;
		if (sdata == 0 ||
		    sdata == ((mode & (1 << i)) ? 0xffff : 0xff))
			edata->state = ENVSYS_SINVALID;
		else {
			edata->value_cur = 1350000 / 2 / sdata;
			edata->state = ENVSYS_SVALID;
		}
		DPRINTF(("%s: sdata[fan%d] 0x%x div: 0x%x\n", __func__,
		    i, sdata, divisor));
	}
#undef FANDATA
}

static int              
it_gtredata(struct sysmon_envsys *sme, struct envsys_data *edata)
{
	struct it_softc *sc = sme->sme_cookie;

	if (edata->sensor < IT_VOLTSTART_IDX)
		it_refresh_temp(sc, edata);
	else if (edata->sensor >= IT_VOLTSTART_IDX &&
	         edata->sensor < IT_FANSTART_IDX)
		it_refresh_volts(sc, edata);
	else if (edata->sensor >= IT_FANSTART_IDX)
		it_refresh_fans(sc, edata);
        else
		return 0;

	return 0;
}
